### ISSUE TITLE
polysemy-RandomFu is broken on ghc 9 

### ISSUE THREAD

**Issue by funketh** [adamConnerSax/polysemy-Extra#2]
might be similiar to https://github.com/polysemy-research/polysemy-zoo/pull/70
```hs
[3 of 3] Compiling Polysemy.ConstraintAbsorber.MonadRandom
                                 
/home/theo/prj/polysemy-Extra/polysemy-RandomFu/src/Polysemy/ConstraintAbsorber/MonadRandom.hs:22:15: error:
    * Could not deduce: (Member RandomFu r0) 
      Fix:                       
        add (Member RandomFu r0) to the context of
          the type signature     
    * In the first argument of `RandomDict', namely `getRandomPrim'
      In the third argument of `absorbWithSem', namely
        `(RandomDict getRandomPrim)'
      In the expression:         
        absorbWithSem            
          @R.MonadRandom @Action (RandomDict getRandomPrim) (Sub Dict)
   |                             
22 |   (RandomDict getRandomPrim)
   |               ^^^^^^^^^^^^^ 
                                 
/home/theo/prj/polysemy-Extra/polysemy-RandomFu/src/Polysemy/ConstraintAbsorber/MonadRandom.hs:23:8: error:
    * Could not deduce (R.MonadRandom (Action (Sem r) s))
        arising from a use of `Dict'
      from the context: Member RandomFu r
        bound by the type signature for:
                   absorbMonadRandom :: forall (r :: [(* -> *) -> * -> *]) a.
                                        Member RandomFu r =>
                                        (R.MonadRandom (Sem r) => Sem r a) -> Sem r a
        at src/Polysemy/ConstraintAbsorber/MonadRandom.hs:(19,1)-(20,50)
      or from: Reifies s (RandomDict (Sem r0))
        bound by a type expected by the context:
                   Reifies s (RandomDict (Sem r0)) =>
                   Dict (R.MonadRandom (Action (Sem r) s))
        at src/Polysemy/ConstraintAbsorber/MonadRandom.hs:23:8-11
    * In the first argument of `Sub', namely `Dict'
      In the fourth argument of `absorbWithSem', namely `(Sub Dict)'
      In the expression:         
        absorbWithSem            
          @R.MonadRandom @Action (RandomDict getRandomPrim) (Sub Dict)
   |                             
23 |   (Sub Dict)                
   |        ^^^^                 
```
The `stack.yaml` I used:
```yml
compiler: ghc-9.0.1
resolver: nightly-2020-05-07

allow-newer: true

packages:
- ./polysemy-RandomFu

extra-deps:
- atomic-primops-0.8.4
- compact-0.2.0.0
- flexible-defaults-0.0.3
- ghc-tcplugins-extra-0.4.1
- MonadRandom-0.5.3
- random-1.2.0
- random-source-0.3.0.11
- rvar-0.2.0.6
- splitmix-0.1.0.3
- syb-0.7.2.1
- th-abstraction-0.4.2.0
- QuickCheck-2.14.2
# GHC 9 patches (not yet on hackage)
- github: polysemy-research/polysemy
  commit: 09c65bdae9c9ce50a55c7ee2ab9681916f09f4c3
  subdirs: [., polysemy-plugin]
- github: polysemy-research/polysemy-zoo
  commit: a93ec969d7efd397696a5495afe2e4fa0c509d6f
- github: haskell-numerics/random-fu
  commit: 2d38294582ba211df6a5a2b7f7a2c5e71038b0e7
  subdirs: [random-fu]
- github: mokus0/th-extras
  commit: 57a97b4df128eb7b360e8ab9c5759392de8d1659
```

**Comment by adamConnerSax** [comment]
I’ll check into it!

**Comment by adamConnerSax** [comment]
I can't get th-extras to compile, so I can't get to your error.  I'm getting all the same github versions as you specify in your stack.yaml (though via cabal in my case) but I get errors in th-extras:

```haskell
[1 of 1] Compiling Language.Haskell.TH.Extras ( src/Language/Haskell/TH/Extras.hs, dist/build/Language/Haskell/TH/Extras.o, dist/build/Language/Haskell/TH/Extras.dyn_o )

src/Language/Haskell/TH/Extras.hs:49:17: error:
    • Expecting one more argument to ‘TyVarBndr’
      Expected a type, but ‘TyVarBndr’ has kind ‘* -> *’
    • In the type signature: nameOfBinder :: TyVarBndr -> Name
   |
49 | nameOfBinder :: TyVarBndr -> Name
   |                 ^^^^^^^^^

src/Language/Haskell/TH/Extras.hs:58:27: error:
    • Expecting one more argument to ‘TyVarBndr’
      Expected a type, but ‘TyVarBndr’ has kind ‘* -> *’
    • In the type signature: varsBoundInCon :: Con -> [TyVarBndr]
   |
58 | varsBoundInCon :: Con -> [TyVarBndr]
   |                           ^^^^^^^^^

src/Language/Haskell/TH/Extras.hs:254:28: error:
    • Expecting one more argument to ‘TyVarBndr’
      Expected a type, but ‘TyVarBndr’ has kind ‘* -> *’
    • In the first argument of ‘Q’, namely ‘([TyVarBndr], Int)’
      In the type signature: tyConArity' :: Name -> Q ([TyVarBndr], Int)
    |
254 | tyConArity' :: Name -> Q ([TyVarBndr], Int)
    |                            ^^^^^^^^^
cabal: Failed to build th-extras-0.0.0.4 (which is required by
polysemy-RandomFu-0.4.2.0). See the build log above for details.
```

**Comment by adamConnerSax** [comment]
Nevermind.  Got past that. Now stuck on compiling polysemy-plugin.

**Comment by funketh** [comment]
well polysemy-plugin should work with the same commit as polysemy (09c65bdae9c9ce50a55c7ee2ab9681916f09f4c3)

**Comment by adamConnerSax** [comment]
Okay! Got past that and am now using your random-fu (https://github.com/funketh/random-fu/commits/master) to finally get to the error you pointed out.  I'll see what I can do!

**Comment by adamConnerSax** [comment]
This turned out to be a template haskell issue with where the splice for the MonadRandom instance was in the file.  Should work now, if you use: https://github.com/adamConnerSax/polysemy-Extra/commit/bc110fcd4abe3d8f0cd65ec7b66f285eddf71ad0

@funketh, Can you double-check that this works for you?  Then I will merge and release, since this also works fine for earlier versions of ghc.  Thanks for catching this and fixing the first error!

**Comment by funketh** [comment]
yup, works fine :+1:

**Comment by adamConnerSax** [comment]
Fixed in ee7fee7bac207286c8763b220ddc9058f9495e64
